### ISSUE TITLE
[Bug] Source line information lost for ops eliminated during compilation — many kernel lines unbreakable in msdebug

### ISSUE THREAD

**Issue by MrBambastik** [triton-lang/triton-ascend#323]
## Summary

Many Triton ops are eliminated, folded, absorbed, or otherwise consumed during the compilation pipeline. Each elimination silently drops the op's source `loc(...)`, so the corresponding kernel line never reaches the `.npubin`'s DWARF line table. The result: msdebug cannot set breakpoints on a significant fraction of source lines in a typical kernel, with no indication to the user that the line was elided rather than truly empty.

```
(msdebug) b kernel.py:6
No locations (pending breakpoint).
```

The user has no way to tell, by looking at their kernel, whether a missing breakpoint means "this line really does nothing" or "the compiler ate the carrier for this line."

## Affected categories

We hit this systematically across multiple op classes. Each category loses lines for a different reason:

1. **Replaced ops** — `tt.get_program_id`, `tt.get_num_programs` and similar are replaced with direct kernel-argument reads. The original op is erased; its loc goes with it.

2. **Folded ops in the Triton dialect** — `SplatOp::fold` and `BroadcastOp::fold` collapse splats/broadcasts of constants into `arith.constant dense<...>` *before* any Ascend pass runs. The fold's materialized constant carries the consumer's loc, not the folded op's.

3. **Absorbed pointer arithmetic** — `tt.addptr`'s offset (e.g. `block_start + arange(0, N)`, or 2D `row[:,None]*BLOCK + col[None,:]`) is consumed analytically by `BlockPtrAnalysis` and rolled into `memref.reinterpret_cast` strides/offsets. The defining ops (`arith.muli`, `arith.addi`, `tt.make_range`, `tt.expand_dims`) are erased before any converter sees them.

4. **Layout-only ops that survive but are optimized away later** — `tt.trans`, `tt.reshape`, `tt.expand_dims` lower to `linalg.transpose`/`tensor.reshape`/etc. in TTAdapter, then bishengir eliminates them downstream as zero-instruction transforms.

5. **Callsite-wrapped library ops** — `tl.sum` and other helpers from `triton/language/standard.py` produce `tt.reduce` whose loc is `callsite(standard.py:N at kernel.py:M)`. DWARF emission resolves this to the library frame (standard.py), not the kernel frame, so the kernel line is missing while a useless library line appears instead.

6. **Fused locations** — when two source-level ops collapse into one op (e.g. `tl.full` + `tl.broadcast_to` → single `linalg.fill`), the survivor carries a `FusedLoc[line_A, line_B]`. bishengir's DWARF emitter collapses `FusedLoc` to the function-level loc, losing both inner lines.

## Reproducer

A kernel exercising all the categories above:

```python
import torch, triton
import triton.language as tl

@triton.jit
def k(xp, op4, op1, n, BLOCK: tl.constexpr, ROWS: tl.constexpr):
    pid  = tl.program_id(0)                                 # category 1
    np_  = tl.num_programs(0)                               # category 1
    offs = pid * BLOCK + tl.arange(0, BLOCK)                # category 3 (1D)
    one  = tl.full((BLOCK,), 1.0, dtype=tl.float32)         # category 2
    x    = tl.load(xp + offs)
    bcast = tl.broadcast_to(tl.reshape(x, (1, BLOCK)), (ROWS, BLOCK))   # category 4
    trans = tl.trans(tl.reshape(x, (1, BLOCK)))             # category 4
    row   = tl.arange(0, ROWS)
    offs2 = row[:, None] * BLOCK + tl.arange(0, BLOCK)[None, :]   # category 3 (2D)
    s     = tl.sum(bcast.to(tl.float32), axis=1)            # category 5
    tl.store(op1 + offs,  x + one + tl.sum(trans.to(tl.float32), axis=1)[:, 0])
    tl.store(op4 + offs2, bcast + s[:, None])
```

```bash
msdebug python3 kernel.py
(msdebug) b kernel.py:6
(msdebug) r
```

Compile, then:

```bash
readelf --debug-dump=decodedline cache/<HASH>/k.npubin | grep <kernel.py>
```

Compare against the lines present in the TTIR `loc(...)` references. The difference is the lines that vanished — at the time of this report, ~8 distinct categories of loss were reproducible with kernels along these lines.

## Why it matters

- Source-level debugging is the primary way users diagnose kernel issues, especially perf and correctness bugs that only show up on real hardware (where `TRITON_INTERPRET` doesn't help).
- The loss is silent: no warning, no indication. Users blame their setup or their syntax when in fact the compiler ate the line.
- The pattern of loss is non-obvious: visually "trivial" lines (a `tl.trans`, a `tl.full`) seem like they should work, but don't, while "complex" lines (loads, atomics) tend to work fine. The mental model is exactly inverted from where users expect breakpoints to land.

## Current workaround (PR #151)

When `TRITON_DEBUG=1` is set, the Ascend pipeline emits `llvm.inline_asm "nop"` markers tagged with the original op's loc, before that op is replaced/absorbed/erased. The NOPs become real PCs in the `.npubin`, each carrying a DWARF entry for its source line. With `TRITON_DEBUG` unset, no NOPs are emitted and the binary is bit-identical to the pre-patch output.

### Environment details

- triton-ascend `release/3.5.x`
- 910B4
- LLVM/MLIR `fad32722`
- TRITON_DISABLE_FFTS=0
- TRITON_DEBUG=1
- TRITON_DISABLE_LINE_INFO=0
- TRITON_KERNEL_DUMP=1
- TRITON_ALWAYS_COMPILE=1
- TRITON_DEVICE_PRINT=1

**Comment by WuTYSFG** [comment]
Thank you for your contribution. We will arrange for a committer to review it.

**Comment by MrBambastik** [comment]
fixed by PR [1151](https://github.com/triton-lang/triton-ascend/pull/1151)
